vscode/src/copilot/webview/copilot.tsx

```diff
@@ -0,0 +1,382 @@
+      </div>
+      <div className="input-box-divider" ref={hrRef} />
+    </>
+  );
+}
+
+/**
+ * A user message in the chat history.
+ */
+function UserMessage(props: { content: string }) {
+  return (
+    <div className="right-message-row">
+      <div className="user-message">
+        <Markdown markdown={props.content} />
+      </div>
+    </div>
+  );
+}
+
+/**
+ * An assistant message in the chat history.
+ */
+function AssistantMessage(props: { content: string }) {
+  return (
+    <div className="left-message-row">
+      <div className="assistant-message">
+        <Markdown markdown={props.content}></Markdown>
+      </div>
+    </div>
+  );
+}
+
+/**
+ * A widget in the chat history.
+ *
+ * A widget "message" is not a real message in the chat: it's
+ * inserted into the chat by tool calls, and displayed
+ * in the client, but it's not included in the chat history
+ * payload that gets sent to the service.
+ */
+function Widget(props: { histogram: HistogramData }) {
+  const histoMap: Map<string, number> = new Map(props.histogram.buckets);
+  return (
+    <div className="left-message-row">
+      <div className="histogram-container">
+        <Histogram
+          data={histoMap}
+          filter=""
+          shotCount={props.histogram.shotCount}
+          onFilter={() => undefined}
+          shotsHeader={false}
+        />
+      </div>
+    </div>
+  );
+}
+
+type ChatViewModel = {
+  /**
+   * The complete chat history.
+   */
+  history: ChatElement[];
+
+  /**
+   * The current status, e.g. "ready", "waitingForAssistantResponse", etc.
+   */
+  status: Status;
+
+  /**
+   * Service backend options.
+   */
+  serviceOptions: ServiceType[];
+
+  /**
+   * Service backend in use.
+   */
+  service?: ServiceType;
+
+  /**
+   * Any in progress assistant response.
+   */
+  inProgressResponse: string;
+};
+
+/**
+ * Copilot command to add a user message to the
+ * current chat and request a response.
+ */
+function submitUserMessage(content: string) {
+  postMessageToExtension({
+    command: "submitUserMessage",
+    request: content,
+  });
+}
+
+/**
+ * Copilot command to restart the chat with a new history.
+ * The service backend can be changed here as well.
+ */
+function restartChat(history: ChatElement[], service?: ServiceType) {
+  postMessageToExtension({
+    command: "restartChat",
+    history,
+    service,
+  });
+}
+
+/**
+ * Holds the current UI state.
+ */
+const model: ChatViewModel = {
+  history: [],
+  inProgressResponse: "",
+  status: { status: "ready" },
+  service: undefined,
+  serviceOptions: [],
+};
+
+document.addEventListener("DOMContentLoaded", loaded);
+window.addEventListener("message", onMessage);
+
+function loaded() {
+  console.log("Loaded");
+  restartChat([], undefined);
+}
+
+/**
+ * Handles updates to the copilot state.
+ */
+function onMessage(event: MessageEvent<CopilotUpdate>) {
+  console.log("Received message: ", event.data);
```

Comment:
Do we want to console.log all messages received by default?

Reply:
Nope I should remove this.